eth/filters/filter_system.go

```diff
@@ -14,172 +14,588 @@
+		for {
+			select {
+			case logs := <-f.logs:
+				allLogs = append(allLogs, logs...)
+			default: // available logs read
+				return allLogs, nil
 			}
-			fs.filterMu.RUnlock()
-		case core.TxPreEvent:
-			fs.filterMu.RLock()
-			for _, filter := range fs.pendingTxFilters {
-				if filter.TransactionCallback != nil && !filter.created.After(event.Time) {
-					filter.TransactionCallback(ev.Tx)
-				}
+		}
+	}
+
+	return nil, errFilterNotFound
+}
+
+// GetLogFilterChanges returns all logs matching the criteria for the filter with the given filter id.
+func (m *Manager) GetLogFilterChanges(id FilterID) ([]Log, error) {
+	m.allMu.RLock()
+	defer m.allMu.RUnlock()
+
+	if f, found := m.all[id]; found && f.typ == LogFilter {
+		f.lastUsed = time.Now()
+		allLogs := make([]Log, 0, len(f.logs)) // prevent (most) allocs for the append
+		for {
+			select {
+			case logs := <-f.logs:
+				allLogs = append(allLogs, logs...)
+			default: // available logs read
+				return allLogs, nil
 			}
-			fs.filterMu.RUnlock()
-
-		case vm.Logs:
-			fs.filterMu.RLock()
-			for _, filter := range fs.logFilters {
-				if filter.LogCallback != nil && !filter.created.After(event.Time) {
-					for _, log := range filter.FilterLogs(ev) {
-						filter.LogCallback(log, false)
-					}
-				}
+		}
+	}
+
+	return nil, errFilterNotFound
+}
+
+// NewPendingTransactionFilter creates a filter that retrieves pending transactions.
+func (m *Manager) NewPendingTransactionFilter() (FilterID, error) {
+	id, err := newID()
+	if err != nil {
+		return FilterID{}, err
+	}
+
+	f := &filter{
+		ID:         id,
+		created:    time.Now(),
+		lastUsed:   time.Now(),
+		canTimeout: true,
+		typ:        PendingTxFilter,
+		hashes:     make(chan common.Hash, maxPendingHashes),
+		logs:       make(chan []Log, maxPendingLogs),
+	}
+
+	m.allMu.Lock()
+	m.all[id] = f
+	m.allMu.Unlock()
+
+	m.install <- f
+
+	return id, nil
+}
+
+// GetPendingTxFilterChanges returns hashes for pending transactions which are added since the last poll.
+func (m *Manager) GetPendingTxFilterChanges(id FilterID) ([]common.Hash, error) {
+	m.allMu.RLock()
+	defer m.allMu.RUnlock()
+
+	if f, found := m.all[id]; found && f.typ == PendingTxFilter {
+		f.lastUsed = time.Now()
+		hashes := make([]common.Hash, 0, len(f.hashes)) // prevent (most) allocs
+		for {
+			select {
+			case hash := <-f.hashes:
+				hashes = append(hashes, hash)
+			default: // read available tx hashes
+				return hashes, nil
+			}
+		}
+	}
+
+	return nil, errFilterNotFound
+}
+
+type filterIndex map[Type]map[FilterID]*filter
+
+// process an event and forward to filters that match the criteria.
+func (m *Manager) process(filters filterIndex, ev *event.Event) {
+	var inactive []*filter
+
+	logHandler := func(f *filter, logs []Log) {
+		if f.lc != nil && len(logs) > 0 {
+			f.lc(f.ID, logs)
+		} else if len(logs) > 0 {
+			select {
+			case f.logs <- logs:
+				return
+			default: // data queue full, disable filter
+				inactive = append(inactive, f)
```

Comment:
The queue is used for polling, subscriptions use the callback which doesn't cause a problem. 

Consider the following case if we block or use an uninstall channel:
1. A DApp installs a block filter over http (polling)
2. The node start synchronising and generates a lot of these block events
3. The DApp doesn't query the filter for changes
4. The filter hash queue is full at some point, without the default case this will block until the filter is polled. We cannot expect the DApp to call the Uninstall method.